### PR TITLE
feat: refactor access control to enable swapping implementation

### PR DESCRIPTION
**Pull request by khorshuheng**
- Consolidate all casbin specific implementation to an optional mod in access-control crate. This simplifies the process of replacing access control implementation with another (eg. external authorization service instead of casbin, or a simpler implementation that is purely based on postgres).
- Remove is_access_control_enabled from the access control crate, for better separation of concern. Whether access control should be enabled or not should be the responsibility of the caller, not the implementation of access control itself.
- Having a configuration that enable or disable access control within the implementation also complicates the implementation since we have to make sure that every single method required by the trait does not perform access control when access control is disabled.
- Use dynamic dispatch as opposed to generics, so that we can switch implementation of access control at run time (eg. use NoOps access control is access control is disabled).

### CODE REVIEW

File: libs/access-control/src/casbin/access.rs [hunk 1]
```diff
@@ -80,30 +80,22 @@ impl AccessControl {
...
-      self.enforcer.remove_policy(uid, obj).await?;
-      let _ = self.change_tx.send(AccessControlChange::RemovePolicy {
-        uid: *uid,
-        oid: obj.object_id().to_string(),
-      });
-      Ok(())
-    } else {
-      Ok(())
-    }
+    self.enforcer.remove_policy(uid, obj).await?;
```
Comment (appflowy) [inline libs/access-control/src/casbin/access.rs:93]: would it be better if `update_policy` and `remove_policy` return an option value. Then change_tx will send if the change is not `None`

File: libs/access-control/Cargo.toml [hunk 2]
```diff
@@ -8,7 +8,11 @@ actix-http.workspace = true
 app-error.workspace = true
 anyhow.workspace = true
 async-trait.workspace = true
-casbin = { version = "2.2.0", features = ["cached", "runtime-tokio", "incremental"] }
+casbin = { version = "2.2.0", features = [
```
Comment (appflowy) [inline libs/access-control/Cargo.toml:11]: we can try to use 2.5.0 version

File: libs/access-control/src/noops/collab.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+use async_trait::async_trait;
+use database_entity::dto::AFAccessLevel;
+
+use crate::{
+  act::Action,
+  collab::{CollabAccessControl, RealtimeAccessControl},
+};
+
+#[derive(Clone)]
+pub struct CollabAccessControlImpl {}
```
Comment (appflowy) [inline libs/access-control/src/noops/collab.rs:11]: can use use `pub struct CollabAccessControlImpl;`

File: libs/access-control/src/noops/collab.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    Ok(())
+  }
+
+  async fn remove_access_level(&self, _uid: &i64, _oid: &str) -> Result<(), AppError> {
+    Ok(())
+  }
+}
+
+#[derive(Clone)]
+pub struct RealtimeCollabAccessControlImpl {}
```
Comment (appflowy) [inline libs/access-control/src/noops/collab.rs:62]: save as abve

File: libs/access-control/src/noops/workspace.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+impl Default for WorkspaceAccessControlImpl {
+  fn default() -> Self {
+    Self::new()
+  }
+}
+
+#[async_trait]
+impl WorkspaceAccessControl for WorkspaceAccessControlImpl {
```
Comment (appflowy) [inline libs/access-control/src/noops/workspace.rs:25]: Not sure we should provide all default implement in AccessControl traits. Then we don't need to provide the default implementation its `impl`

File: libs/access-control/Cargo.toml (hunk 2 above)
Comment (khorshuheng) [reply libs/access-control/Cargo.toml:11]: Just tried using the 2.5.0 version - there are some changes in function signature. Will do the upgrade in another PR instead and keep 2.2.0 for this PR.

**Pull request merged**